barretenberg/cpp/src/barretenberg/ecc/scalar_multiplication/sorted_msm.cpp

```diff
@@ -0,0 +1,240 @@
+ * @details Addition of points P_1, P_2 requires computation of a term of the form 1/(P_2.x - P_1.x). For efficiency,
+ * these terms are computed all at once for a full set of addition sequences using batch inversion.
+ *
+ * @tparam Curve
+ * @param add_sequences
+ */
+template <typename Curve>
+void MsmSorter<Curve>::batch_compute_point_addition_slope_inverses(AdditionSequences& add_sequences)
+{
+    auto points = add_sequences.points;
+    auto sequence_counts = add_sequences.sequence_counts;
+
+    // Count the total number of point pairs to be added across all addition sequences
+    size_t total_num_pairs{ 0 };
+    for (auto& count : sequence_counts) {
+        total_num_pairs += count >> 1;
+    }
+
+    // Define scratch space for batched inverse computations and eventual storage of denominators
+    std::span<Fq> scratch_space(denominators.data(), total_num_pairs);
+    std::vector<Fq> differences;
+    differences.resize(total_num_pairs);
+
+    // Compute and store successive products of differences (x_2 - x_1)
+    Fq accumulator = 1;
+    size_t point_idx = 0;
+    size_t pair_idx = 0;
+    for (auto& count : sequence_counts) {
+        const auto num_pairs = count >> 1;
+        for (size_t j = 0; j < num_pairs; ++j) {
+            const auto& x1 = points[point_idx++].x;
+            const auto& x2 = points[point_idx++].x;
+
+            // It is assumed that the input points are random and thus w/h/p do not share an x-coordinate
+            ASSERT(x1 != x2);
+
+            auto diff = x2 - x1;
+            differences[pair_idx] = diff;
+
+            // Store and update the running product of differences at each stage
+            scratch_space[pair_idx++] = accumulator;
+            accumulator *= diff;
+        }
+        // If number of points in the sequence is odd, we skip the last one since it has no pair
+        point_idx += (count & 0x01ULL);
+    }
+
+    // Invert the full product of differences
+    Fq inverse = accumulator.invert();
+
+    // Compute the individual point-pair addition denominators 1/(x2 - x1)
+    for (size_t i = 0; i < total_num_pairs; ++i) {
+        size_t idx = total_num_pairs - 1 - i;
+        scratch_space[idx] *= inverse;
+        inverse *= differences[idx];
+    }
+}
+
+/**
+ * @brief Add two affine elements with the inverse in the slope term \lambda provided as input
+ * @details The sum of two points (x1, y1), (x2, y2) is given by x3 = \lambda^2 - x1 - x2, y3 = \lambda*(x1 - x3) - y1,
+ * where \lambda  = (y2 - y1)/(x2 - x1). When performing many additions at once, it is more efficient to batch compute
+ * the inverse component of \lambda for each pair of points. This gives rise to the need for a method like this one.
+ *
+ * @tparam Curve
+ * @param point_1 (x1, y1)
+ * @param point_2 (x2, y2)
+ * @param denominator 1/(x2 - x1)
+ * @return Curve::AffineElement
+ */
+template <typename Curve>
+typename Curve::AffineElement MsmSorter<Curve>::affine_add_with_denominator(const G1& point_1,
```

Comment:
Since we call this all the time in one loop it would be helpful to inline it, I think

Reply:
Thanks, done